src/dns.h

```diff
@@ -60,40 +66,61 @@ typedef struct DNSPacket
-static int	        destroy_DNSQuery(DNSQuery** query);
-static int          destroy_DNSQuestion(DNSQuestion** question);
-static int          destroy_DNSRecord(DNSRecord** record);
+/* -------------------------------------------
+ * DNS Resolution
+ * ------------------------------------------- */
+void                resolve(char* domain_name, uint16_t record_type, char answer[]);
+static DNSPacket *  send_query(char *addr, char *domain, u_int16_t type);
+static int          get_answer(DNSPacket* packet, char buf[]);
+static int          get_nameserver_ip(DNSPacket* packet, char buf[]);
+static int          get_nameserver(DNSPacket* packet, char buf[]);
 
+/* -------------------------------------------
+ * DNS Header
+ * ------------------------------------------- */
+static DNSHeader*   NewDNSHeader(uint16_t id, uint16_t flags, uint16_t num_questions);
+static int          destroy_DNSHeader(DNSHeader** header);
 static void         display_DNSHeader(DNSHeader *header);
-static void         display_DNSQuestion(DNSQuestion *question);
-static void         display_DNSRecord(DNSRecord *record);
-static void         display_DNSPacket(DNSPacket *packet);
-
-static size_t       encode_dns_name(char* domain_name, char* res);
 static void         header_to_bytes(DNSHeader *header, char *header_bytes);
-static void         question_to_bytes(DNSQuestion *question, char *question_bytes);
 static size_t       parse_header(const unsigned char* response_bytes, DNSHeader *header);
+
+/* -------------------------------------------
+ * DNS Question
+ * ------------------------------------------- */
+static DNSQuestion* NewDNSQuestion(char *encoded_name, uint16_t type, uint16_t class);
+static int          destroy_DNSQuestion(DNSQuestion** question);
+static void         display_DNSQuestion(DNSQuestion *question);
+static void         question_to_bytes(DNSQuestion *question, char *question_bytes);
 static int          parse_questions(const unsigned char *response_bytes, int bytes_read, int num_questions, DNSQuestion **head);
 static int          parse_question(const unsigned char* response_bytes, int bytes_in, DNSQuestion *question);
+
+/* -------------------------------------------
+ * DNS Query
+ * ------------------------------------------- */
+static DNSQuery*    NewDNSQuery(char *domain_name, uint16_t record_type);
+static int	        destroy_DNSQuery(DNSQuery** query);
+
+/* -------------------------------------------
+ * DNS Packet
+ * ------------------------------------------- */
+static DNSPacket*   NewDNSPacket(const unsigned char *response_bytes);
+static int          destroy_DNSPacket(DNSPacket** packet);
+static void         display_DNSPacket(DNSPacket *packet);
+
+/* -------------------------------------------
+ * DNS Record
+ * ------------------------------------------- */
+static int          destroy_DNSRecord(DNSRecord** record);
+static void         display_DNSRecord(DNSRecord *record);
```

Comment:
From a structure point of view, Packet = Header + Questions + Records, so the DNS Packet functions could be placed after DNS Record